lib/BaseDownloadController.js

```diff
@@ -77,57 +81,141 @@ module.exports = class BaseDownloadController extends CompositeController {
-          if (Array.isArray(file) && file.length == 1) { file = file[0]; }
-
-          // TODO if last-modified doesnt exist try etag
-          // use with request at .headers.If-None-Match
-          let resLM = objectPath.get(res, 'headers.last-modified');
-          if (resLM) {
-            await fs.outputJson(fileCachePath, file);
-            newLastModifiedArray[i] = { hash: requestHash, url: url, 'last-modified': resLM };
+      const git = objectPath.get(req, 'options.git');
+      let host;
+      let repo;
+      let branch;
+      let requrl;
+      let files;
+      let path;
+      let fileExt;
+      let filename;
+      if (git) {
+        const parse = gh(git.repo);
+        host = parse.host;
+        repo = parse.repo;
+        branch = git.branch;
+        const pattern = parsePath(git.filePath);
+        if (pattern.ext == '') {
+          if (git.filePath.charAt(git.filePath.length-1) == '*') {
+            path = pattern.dir;
+          } else {
+            path = pattern.path;
           }
-        } else if (res.statusCode == 304 && fileCached) {
-          this.log.debug(`Download ${res.statusCode} Not Modified ${url}`);
-          file = await fs.readJson(fileCachePath);
-          newLastModifiedArray[i] = { hash: requestHash, url: url, 'last-modified': objectPath.get(imsObj, 'last-modified') };
+          fileExt = pattern.ext;
         } else {
-          this.log.debug(`Download failed: ${res.statusCode} | ${url}`);
-          throw { statusCode: res.statusCode, uri: url };
+          if (pattern.stem == '*') {
+            fileExt = pattern.ext;
+          } else {
+            filename = pattern.base;
+          }
+          path = pattern.dir;
         }
-      } catch (e) {
-        if (optional) {
-          ++optionalResourceFailure;
-          this.log.warn(e.message || e);
-          this.updateRazeeLogs('warn', { controller: 'BaseDownload', warn: `Error downloading file. StatusCode: ${e.statusCode}`, url: url });
-          file = [];
-          newLastModifiedArray[i] = { hash: requestHash, url: url };
-        } else {
-          if (e.message === undefined) this.log.error(e);
-          return Promise.reject(`uri: ${reqOpt.uri || reqOpt.url}, statusCode: ${e.statusCode}, message: ${e.message}`);
+
+        if (git.provider == 'github') {
+          let enterprise = '';
+          if (host == 'github.ibm.com') {
+            enterprise = 'http://github.ibm.com/api/v3';
+          }
+          if (path.charAt(path.length-1) == '/') {
+            path = path.substring(0, path.length-1);
+          }
+          requrl = `GET ${enterprise}/repos/${repo}/contents/${path}?ref=${branch}`;
+          if (reqOpt.headers.Authorization && !reqOpt.headers.Authorization.includes('token')) {
+            reqOpt.headers = { ...reqOpt.headers, Authorization: 'token ' + reqOpt.headers.Authorization };
+          }
+          
+        } 
+        else if (git.provider == 'gitlab') {
```

Comment:
Consider factoring the git provider logic out into a Factory to make it easier to later add additional providers, while making this method simpler

Reply:
yes! should use a factory like we did for remoteresources3. https://github.com/razee-io/RemoteResource/blob/master/src/RemoteResourceS3Controller.js